File: backend/app/schemas/feihualing.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from uuid import UUID

class FeiHuaLingGameStart(BaseModel):
    difficulty: Optional[int] = 10
    keyword: Optional[str] = None

class FeiHuaLingGameResponse(BaseModel):
    game_id: UUID
    keyword: str
    time_limit: int = 30
    target_rounds: int = 10
    ai_score: int = 0
    ai_first_round: Optional['AIRoundData'] = None

class PoemSubmit(BaseModel):
    game_id: UUID
    poem_content: str
    response_time: int
    target_rounds: Optional[int] = 10

class AIResponse(BaseModel):
    poem_content: str
    author: Optional[str] = None
    title: Optional[str] = None
    dynasty: Optional[str] = None

class AIRoundData(BaseModel):
    verse: str
    poem_title: Optional[str] = None
    author: Optional[str] = None
    dynasty: Optional[str] = None
    round_number: int = 0
    verified: bool = False

class PoemSubmitResponse(BaseModel):
    valid: bool
    message: str
    continue_game: bool
    user_score: int
    ai_score: int = 0
    round_number: int
    score_gained: int = 0
    combo: int = 0
    user_round_count: int = 0
    poem_author: Optional[str] = None
    poem_title: Optional[str] = None
    poem_dynasty: Optional[str] = None
    ai_round: Optional[AIRoundData] = None
    ai_failed: bool = False

class GameEndRequest(BaseModel):
    game_id: UUID
    reason: str

class HintResponse(BaseModel):
    hint: str
    author: str
    hint_cost: int = 0

class RoundHistory(BaseModel):
    round_number: int
    player: str
    poem_content: str
    author: Optional[str] = None
    title: Optional[str] = None
    dynasty: Optional[str] = None
    created_at: datetime

class GameResult(BaseModel):
    game_id: UUID
    result: str
    user_score: int
    ai_score: int = 0
    total_rounds: int
    user_round_count: int = 0
    keyword: str
    duration: int
    rounds: List[RoundHistory]

class GameHistoryItem(BaseModel):
    id: UUID
    keyword: str
    result: str
    user_score: int
    ai_score: int = 0
    total_rounds: int
    difficulty: Optional[int] = 10
    created_at: datetime

    class Config:
        from_attributes = True

class GameHistoryResponse(BaseModel):
    items: List[GameHistoryItem]
    total: int
    page: int
    page_size: int

File: frontend/src/api/feihualing.ts
import request from '@/utils/request'

export interface AIRoundData {
  verse: string
  poem_title?: string
  author?: string
  dynasty?: string
  round_number: number
  verified?: boolean
}

export interface FeiHuaLingGameResponse {
  game_id: string
  keyword: string
  time_limit: number
  target_rounds: number
  ai_score: number
  ai_first_round?: AIRoundData
}

export interface PoemSubmitRequest {
  game_id: string
  poem_content: string
  response_time: number
  target_rounds?: number
}

export interface PoemSubmitResponse {
  valid: boolean
  message: string
  continue_game: boolean
  user_score: number
  ai_score: number
  round_number: number
  score_gained: number
  combo: number
  user_round_count: number
  poem_author?: string
  poem_title?: string
  poem_dynasty?: string
  ai_round?: AIRoundData
  ai_failed?: boolean
  retry?: boolean
}

export interface GameEndRequest {
  game_id: string
  reason: string
}

export interface RoundHistory {
  round_number: number
  player: string
  poem_content: string
  author?: string
  title?: string
  dynasty?: string
  created_at: string
}

export interface GameResult {
  game_id: string
  result: string
  user_score: number
  ai_score: number
  total_rounds: number
  user_round_count: number
  keyword: string
  duration: number
  rounds: RoundHistory[]
}

export interface GameHistoryItem {
  id: string
  keyword: string
  result: string
  user_score: number
  ai_score: number
  total_rounds: number
  created_at: string
}

export interface GameHistoryResponse {
  items: GameHistoryItem[]
  total: number
  page: number
  page_size: number
}

export const feiHuaLingApi = {
  startGame(difficulty?: number, keyword?: string): Promise<FeiHuaLingGameResponse> {
    return request.post('/feihualing/start', { difficulty, keyword }, { timeout: 30000 })
  },

  submitPoem(data: PoemSubmitRequest): Promise<PoemSubmitResponse> {
    return request.post('/feihualing/submit', data, { timeout: 15000 })
  },

  getHint(gameId: string): Promise<{ hint: string; author: string; hint_cost: number }> {
    return request.get(`/feihualing/hint/${gameId}`)
  },

  endGame(data: GameEndRequest): Promise<GameResult> {
    return request.post('/feihualing/end', data)
  },

  getHistory(params: { page: number; page_size: number }): Promise<GameHistoryResponse> {
    return request.get('/feihualing/history', { params })
  }
}
